gems/aws-crt-auth/lib/aws-crt-auth/credentials.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Aws
+  module Crt
+    module Auth
+      # Utility class for Credentials.
+      # @api private
+      class Credentials
+        UINT64_MAX = 18_446_744_073_709_551_615
+
+        # @param [String] access_key_id
+        # @param [String] secret_access_key
+        # @param [String] session_token (nil)
+        # @param [Time|int] expiration (nil)
+        def initialize(access_key_id, secret_access_key,
+                       session_token = nil, expiration = nil)
+          if !access_key_id || access_key_id.empty?
+            raise ArgumentError, 'access_key_id  must be set'
+          end
+
+          if !secret_access_key || secret_access_key.empty?
+            raise ArgumentError, 'secret_access_key  must be set'
+          end
+
+          native = Aws::Crt.call do
+            Aws::Crt::Native.credentials_new(
+              access_key_id,
+              secret_access_key,
+              session_token,
+              expiration&.to_i || UINT64_MAX
+            )
+          end
+
+          @native = FFI::AutoPointer.new(native, self.class.method(:on_release))
+        end
+
+        # @return [FFI:Pointer]
+        attr_reader :native
+
+        # @return [String, nil]
+        def access_key_id
+          Aws::Crt::Native.credentials_get_access_key_id(@native) if @native
```

Comment:
for `def access_key_id` and `def secret_access_key` it should not be possible for these to return nil. These are required args.

If `@native` is nil here because the pointer has been freed, that should result in an exception. For **every** member function (except maybe release()), @native being nil should result in an exception.

Reply:
Yeah - I think this is a good point.  There is a lot of generic `native` wrapper functionality that I'd like to generalize into a module that can just be included in these classes to give us the release/on_release + some utilities for checking/raising those exceptions.